activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
> I think this will create the duplicate of value whenever we are not passing :for key

Doesn't the current implementation do that?

``` ruby
elsif value.is_a?(Array)
  unless options[:for] == :assignment
    value = value.dup
  end
  value.map! { |e| convert_value(e, options) }
else
```

That dups it if `:for` is anything other than `:assignment` including nil.